linux/src/completion_port_linux.c

```diff
@@ -0,0 +1,434 @@
+                    if (epoll_ctl(completion_port->epoll, EPOLL_CTL_MOD, epoll_thread_data->socket, &ev) < 0)
+                    {
+                        if (errno == ENOENT)
+                        {
+                            // Codes_COMPLETION_PORT_LINUX_11_024: [ If the epoll_ctl call fails with ENOENT, completion_port_add shall call epoll_ctl again with EPOLL_CTL_ADD. ]
+                            if (epoll_ctl(completion_port->epoll, EPOLL_CTL_ADD, epoll_thread_data->socket, &ev) < 0)
+                            {
+                                LogErrorNo("failure with epoll_ctl EPOLL_CTL_ADD");
+                            }
+                            else
+                            {
+                                result = 0;
+                                (void)interlocked_decrement(&completion_port->pending_calls);
+                                wake_by_address_single(&completion_port->pending_calls);
+                                goto all_ok;
+                            }
+                        }
+                        else
+                        {
+                            LogErrorNo("failure with epoll_ctl EPOLL_CTL_MOD");
+                        }
+                    }
+                    else
+                    {
+                        // Codes_COMPLETION_PORT_LINUX_11_026: [ On success, completion_port_add shall return 0. ]
+                        result = 0;
+                        // Codes_COMPLETION_PORT_LINUX_11_025: [ completion_port_add shall decrement the ongoing call count value to unblock close. ]
+                        (void)interlocked_decrement(&completion_port->pending_calls);
+                        wake_by_address_single(&completion_port->pending_calls);
+                        goto all_ok;
+                    }
+                    // Remove
+                    remove_thread_data_from_list(completion_port, epoll_thread_data);
+                }
+                free(epoll_thread_data);
+            }
+            // Codes_COMPLETION_PORT_LINUX_11_027: [ If any error occurs, completion_port_add shall fail and return a non-zero value. ]
+            result = MU_FAILURE;
+            (void)interlocked_decrement(&completion_port->pending_calls);
+            wake_by_address_single(&completion_port->pending_calls);
+        }
+    }
+all_ok:
+    return result;
+}
+
+void completion_port_remove(COMPLETION_PORT_HANDLE completion_port, SOCKET_HANDLE socket)
+{
+    if (
+        // Codes_COMPLETION_PORT_LINUX_11_028: [ If completion_port is NULL, completion_port_remove shall return. ]
+        completion_port == NULL || 
+        // Codes_COMPLETION_PORT_LINUX_11_029: [ If socket is INVALID_SOCKET, completion_port_remove shall return. ]
+        socket == INVALID_SOCKET)
+    {
+        LogError("Invalid arguments: COMPLETION_PORT_HANDLE completion_port=%p, SOCKET_HANDLE socket=%" PRI_SOCKET "", completion_port, socket);
+    }
+    else
+    {
```

Comment:
No `pending_calls`? Of course with my other comment I'm wondering if it's needed at all

Reply:
I believe I need to add pending calls here.